File: src/extension.ts
// Composition root for the Energy State Analyzer VS Code extension. This file OWNS activation
// wiring and cross-module coordination only — it creates collaborators, threads shared state (grammar
// caches, decoration set, diagnostics collection) into them, and wires event handlers. It contains no
// detector, presentation, or analysis logic of its own; if you find yourself adding business logic
// here, extract it into a domain module (src/decorations.ts, src/diagnostics.ts, src/grammar.ts,
// src/analysis.ts). Coherence: keep ≤12 functions, ≤10 imports.
import * as vscode from 'vscode';

import { applyDecorations, createDecorations, disposeDecorations, DecorationSet } from './decorations';
import { updateProblemsPanel } from './diagnostics';
import { getOrLoadLanguage, initializeParser, LoadedLanguage } from './grammar';
import { analyzeDocument, isDocumentIgnored } from './analysis';
import { LANGUAGES } from './languages';

// One tree-sitter Parser per supported language, keyed by vscode languageId.
// decision: populated lazily (see getOrLoadLanguage) rather than up front for every
// registered language — grammars range from Python's 448KB to F#'s 12MB, and loading
// all of them at activation makes every window pay that cost even for a single-language
// project
let loadedLanguages: Map<string, LoadedLanguage>;
// Dedupes concurrent loads of the same not-yet-loaded language — onDidChangeTextDocument
// can fire again before the first getOrLoadLanguage call for that language resolves.
let inFlightLoads: Map<string, Promise<LoadedLanguage>>;
let extensionPath: string;

// Create diagnostics collection at module level
let diagnosticsCollection: vscode.DiagnosticCollection;

// The decoration types created once per activation and threaded through the presentation
// functions in ./decorations — this composition root owns them so that module holds no state.
// Assigned synchronously in activate before any event handler can run, mirroring the original
// non-optional module-level decoration bindings.
let decorations: DecorationSet;

export async function activate(context: vscode.ExtensionContext) {
    console.log('🚀 Activating Energy State Analyzer...');

    try {
        // Initialize Parser
        console.log('🔧 Initializing Parser...');
        await initializeParser();
        console.log('✅ Parser initialized');

        loadedLanguages = new Map();
        inFlightLoads = new Map();
        extensionPath = context.extensionPath;

        // Create decoration types
        decorations = createDecorations();
        console.log('🎨 Decoration types created');

        // Create diagnostics collection for Problems panel
        diagnosticsCollection = vscode.languages.createDiagnosticCollection('energyState');
        context.subscriptions.push(diagnosticsCollection);
        console.log('📋 Diagnostics collection created');

        // Register command
        const disposable = vscode.commands.registerCommand('energy-state-analyzer.analyze', () => {
            vscode.window.showInformationMessage('Energy State Analyzer: Manual analysis triggered!');
            void analyzeActiveEditor();
        });
        context.subscriptions.push(disposable);

        // Register event listeners
        vscode.window.onDidChangeActiveTextEditor(() => void analyzeActiveEditor());
        // tradeoff: re-parses and re-runs every detector on every keystroke rather than debouncing — keeps decorations and Problems-panel entries always in sync with the visible buffer, at the cost of re-analysis work the user never sees skipped
        vscode.workspace.onDidChangeTextDocument((event) => {
            if (event.document === vscode.window.activeTextEditor?.document) {
                void analyzeActiveEditor();
            }
        });
        vscode.workspace.onDidChangeConfiguration((event) => {
            if (event.affectsConfiguration('energyStateAnalyzer.colors')) {
                disposeDecorations(decorations);
                decorations = createDecorations();
            }
            if (event.affectsConfiguration('energyStateAnalyzer')) {
                void analyzeActiveEditor();
            }
        });

        // Clear diagnostics when document is closed
        vscode.workspace.onDidCloseTextDocument((document) => {
            if (document.languageId in LANGUAGES) {
                diagnosticsCollection.delete(document.uri);
            }
        });

        // Analyze current editor if open
        void analyzeActiveEditor();

        console.log('✅ Energy State Analyzer activated successfully!');
    } catch (error) {
        console.error('Failed to activate Energy State Analyzer:', error);
        vscode.window.showErrorMessage(`Energy State Analyzer failed to activate: ${error}`);
    }
}

async function analyzeActiveEditor() {
    const editor = vscode.window.activeTextEditor;
    console.log('🔍 Analyzing active editor...');

    if (!editor) {
        console.log('❌ No active editor found');
        return;
    }

    if (isDocumentIgnored(editor.document)) {
        console.log('🚫 Ignored by .esaignore:', editor.document.fileName);
        applyDecorations(editor, decorations, []);
        diagnosticsCollection.delete(editor.document.uri);
        return;
    }

    const loaded = await getOrLoadLanguage(editor.document.languageId, {
        extensionPath,
        loadedLanguages,
        inFlightLoads
    });
    if (!loaded) {
        console.log('⚠️ Unsupported language:', editor.document.languageId);
        // Clear diagnostics for unsupported languages
        diagnosticsCollection.clear();
        return;
    }

    // decision: re-reads the active editor after the await above instead of trusting the
    // `editor` captured before it — the user may have switched tabs while the grammar loaded
    if (vscode.window.activeTextEditor?.document !== editor.document) {
        return;
    }

    console.log(`📄 Analyzing ${loaded.adapter.id} file:`, editor.document.fileName);
    const violations = analyzeDocument(editor.document, loaded);
    console.log('🔍 Found', violations.length, 'energy violations');

    // Apply both visual decorations AND problems panel
    applyDecorations(editor, decorations, violations);
    updateProblemsPanel(diagnosticsCollection, editor.document, violations);
}

export function deactivate() {
    // Clean up decorations AND diagnostics
    if (decorations) {
        disposeDecorations(decorations);
    }
    diagnosticsCollection?.dispose();
}

File: src/Extension/Extension.fs
module Energy.Extension.Extension

open System.Collections.Generic
open System.Threading.Tasks

open Fable.Core

open Energy.Extension.Analysis
open Energy.Extension.Configuration
open Energy.Extension.Decorations
open Energy.Extension.Diagnostics
open Energy.Extension.Grammar
open Energy.Extension.VscodeDiagnostics
open Energy.Extension.VscodeDocument
open Energy.Extension.VscodeHost
open Energy.Extension.VscodeIdentity
open Energy.Extension.VscodeWorkspace
open Energy.Languages.Registry

// Composition root: owns lifecycle state and event wiring only. Detection and presentation stay
// in their domain modules, and the grammar caches are reset for every activation.

type private ExtensionState =
    { Grammar: GrammarContext
      Diagnostics: obj
      mutable Decorations: DecorationSet }

let mutable private state: ExtensionState option = None

let console = JS.console

let private clearIgnored editor current =
    let document = editorDocument editor
    applyDecorations editor current.Decorations []
    deleteDiagnostics current.Diagnostics (documentUri document)

let private isCurrentDocument document =
    match activeTextEditor window with
    | null -> false
    | editor -> sameObject (editorDocument editor) document

// decision: re-reads the active editor after grammar loading because users can change tabs while
// the promise is pending; decorations must never be written onto the newly active document.
let private analyzeActiveEditor () : Task<unit> =
    task {
        console.log ("🔍 Analyzing active editor...")

        match state, activeTextEditor window with
        | _, null -> console.log ("❌ No active editor found")
        | None, _ -> ()
        | Some current, editor ->
            let document = editorDocument editor

            if isDocumentIgnored document then
                console.log ("🚫 Ignored by .esaignore: " + documentFileName document)
                clearIgnored editor current
            else
                let! loaded = getOrLoadLanguage (documentLanguageId document) current.Grammar

                match loaded with
                | None ->
                    console.log ("⚠️ Unsupported language: " + documentLanguageId document)
                    clearDiagnostics current.Diagnostics
                | Some _ when not (isCurrentDocument document) -> ()
                | Some loaded ->
                    console.log ("📄 Analyzing " + loaded.Adapter.Id + " file: " + documentFileName document)
                    let violations = analyzeDocument loaded document
                    console.log ("🔍 Found " + string violations.Length + " energy violations")
                    applyDecorations editor current.Decorations violations
                    updateProblemsPanel current.Diagnostics document violations
    }

let private requestAnalysis () = analyzeActiveEditor () |> ignore

let private subscribeEvents context =
    onDidChangeActiveTextEditor window (fun _ -> requestAnalysis ())
    |> addSubscription context

    onDidChangeTextDocument workspace (fun event ->
        match activeTextEditor window with
        | null -> ()
        | editor when sameObject (textDocumentFromEvent event) (editorDocument editor) -> requestAnalysis ()
        | _ -> ())
    |> addSubscription context

    onDidChangeConfiguration workspace (fun event ->
        match state with
        | Some current when affectsConfiguration event "energyStateAnalyzer.colors" ->
            disposeDecorations current.Decorations
            current.Decorations <- createDecorations (getEnergyColors ())
            requestAnalysis ()
        | Some _ when affectsConfiguration event "energyStateAnalyzer" -> requestAnalysis ()
        | _ -> ())
    |> addSubscription context

    onDidCloseTextDocument workspace (fun document ->
        if Map.containsKey (documentLanguageId document) Energy.Languages.Registry.languages then
            state
            |> Option.iter (fun current -> deleteDiagnostics current.Diagnostics (documentUri document)))
    |> addSubscription context

let activate (context: obj) : Task<unit> =
    task {
        console.log ("🚀 Activating Energy State Analyzer...")

        try
            console.log ("🔧 Initializing Parser...")
            do! initializeParser ()
            console.log ("✅ Parser initialized")

            let grammar =
                { ExtensionPath = extensionPath context
                  LoadedLanguages = Dictionary()
                  InFlightLoads = Dictionary() }

            let decorations = createDecorations (getEnergyColors ())
            let diagnostics = createDiagnosticCollection VscodeHost.languages "energyState"

            state <-
                Some
                    { Grammar = grammar
                      Diagnostics = diagnostics
                      Decorations = decorations }

            addSubscription context diagnostics
            console.log ("🎨 Decoration types created")
            console.log ("📋 Diagnostics collection created")

            registerCommand commands "energy-state-analyzer.analyze" (fun () ->
                showInformationMessage window "Energy State Analyzer: Manual analysis triggered!"
                requestAnalysis ())
            |> addSubscription context

            subscribeEvents context
            requestAnalysis ()
            console.log ("✅ Energy State Analyzer activated successfully!")
        with error ->
            console.error ("Failed to activate Energy State Analyzer:", box error)
            showErrorMessage window ("Energy State Analyzer failed to activate: " + string error)
    }

let deactivate () =
    state
    |> Option.iter (fun current ->
        disposeDecorations current.Decorations
        dispose current.Diagnostics)

    state <- None
